### ISSUE TITLE
Add OMEMO encryption  [$100]

### ISSUE THREAD

**Issue by cart** [movim/movim#63]
I haven't seen end to end encryption mentioned anywhere.  For an organization claiming "Our first aim is to protect your privacy" this seems like an important feature.  Have you considered implementing XEP-0027?  Conversations (the android app) supports it.

**Comment by edhelas** [comment]
There's a bounty on BoutySource here https://www.bountysource.com/teams/movim/issues?utm_source=Movim&utm_medium=shield&utm_campaign=bounties_received

Implementing OTR in Movim is really-really tricky because I have to keep the sessions open, even if you reload the pages, something that people do all the time. Other clients don't have this issue because they are running everything in the same "process". So it's definitely not plan yet in Movim. Also OTR has some serious compatibility issues with XMPP (like handling properly the different resources, the history…).

XEP-0027 is Obsolete actually.

**Comment by ghostofkendo** [comment]
It would be great to have end-to-end encryption in Movim.

Although it's somewhat different from Movim, Salut à Toi has OTR support: http://repos.goffi.org/sat/file/8cc7d83141a4/src/plugins/plugin_sec_otr.py
Maybe you could use SàT's implementation as inspiration for Movim?

Also, regarding the difficulty to correctly implement OTR with XMPP, you could perhaps start with a first minimal version without resources support (e.g. an OTR-encrypted conversation is only valid for the resource used to start it), etc.

**Comment by marxistvegan** [comment]
edhelas not to belabor the point but there is also this https://github.com/jappix/jappix/issues/305

**Comment by edhelas** [comment]
Like I said in my previous comment Jappix run in an unique tab so there is no issue to use OTR.js, which is not the case for Movim.

**Comment by edhelas** [comment]
This issue was moved to movim/movim#18

**Comment by mulles** [comment]
There is a quite new encryption "method" designed for instant messaging, currently implemented in conversations android app:
https://en.m.wikipedia.org/wiki/OMEMO_(encryption)

It's worth a lock.

**Comment by dkellner** [comment]
+1 for implementing OMEMO instead of OTR.

**Comment by heyimgay** [comment]
I'd look to [cyph](https://github.com/cyph/cyph) for inspiration, specifically their [websign](https://www.reddit.com/r/encryption/comments/4027ci/how_2_spacex_alums_are_using_encryption_for_good/cyywmc8) solution to guaranteeing trust after the initial visit.

**Comment by edhelas** [comment]
I'm currently having a look at a possible OMEMO implementation in Movim.

**Comment by bb010g** [comment]
Any updates on this or areas assistance would be wanted in?

**Comment by edhelas** [comment]
Just to keep you updated about this ticket. Movim current architecture is making the implementation of OMEMO (or OTR) quite impossible if I want to do it "right" (without storing risky stuff on the server). So I have no plans to do it for now to do it but I leave the ticket open.

**Comment by MRZA-MRZA** [comment]
Do you mean that there is no possibility to store user's keys/fingerptints?

**Comment by mulles** [comment]
Thanks for sharing. May you elaborate what the technical challenges/s is/are. 

Maybe endtoend encryption is meaningless, when you don't have at least one save place which is permanent to store the password/key needed for encryption.
This seems very hard when you are using ANY webbrowser on ANY computer, but the android app may be capable of.
One could use a new key for every session and safely delete those. If futhermore you want to have message history, the only possibility I see is to save the messages unencrypted and only send them encrypted. 

At the end one must say that endtoend is sercurity on top, which is always good, but needs more effort. I say "On top", because if you trust your proper server and the one to whom you send your message the message transfer is already encrypted the whole time when it passes entites you can't trust, as you may not even be able to know them. Thus, this nice decentralized technology can be deployed now, when everybody thinks well when choosing a server.

**Comment by rhz** [comment]
Perhaps end-to-end encryption could be done using PGP like they do in ProtonMail? The idea would be that everything gets stored on the server encrypted and then it's decrypted by the client (the web ui) using javascript and a second password.

**Comment by edhelas** [comment]
To give a small update here. I've made some try and research regarding E2E encryption in Movim and the current architecture definitely doesn't allow it to be done in a "secure" way. Movim is handling the XMPP connection in PHP and deliver a unique session to several tabs/devices at the same time (on most of the others clients everything is done on one device/computer).
So basically I can't handle a OMEMO/OTR session (that should be done on a device) this way.

**Comment by herbsmn** [comment]
Hey @inputmice, just wanted to ping you on this thread so you know about the OMEMO implimentation discussion going on.

**Comment by MRZA-MRZA** [comment]
What are the advantages of single session for all devices?

**Comment by louy2** [comment]
@edhelas So what we need is a client side javascript OMEMO implementation?

**Comment by edhelas** [comment]
@MRZA-MRZA the architecture of Movim allow him to be used as a "simple website" this means that using a unique cookie, you can open several tabs that will be connected to the same session in the backend. This behavior has been extended in 0.8 to share the same session not only between tabs, but also between browsers (basically your Android phone, your desktop Firefox and your Electron desktop app are sharing the same session cookie and then are perfectly synchronized). That's why Movim can be considered as a "light-client" (and Jappix a "heavy-client"), all the processing and session handling is done server side, the browsers are just displaying the result.

This also means that the XMPP session and processing is done server-side. The end-to-end encryption protocols are always designed with the idea that "where the message is send/arrive" = "where the message is encrypted/decrypted". This is not the case with Movim. A copy of the messages are stored, unencrypted, in a database (that way you can reload the page, do pagination… ). If I respect the OMEMO/OTR rules, I could encrypt/decrypt them on the server but this means that they still will be saved and sent in plain text to the browser (with HTTPS on top), so basically it's not end to end anymore.

Now let's say that the encryption layer is done in JS, in the browser. Because here we could have several tabs/browsers opened on the same "Movim session" they will not share the same OMEMO/OTR keys/status. Also if you reload the page I'll have to sync the current status somewhere to be sure that I'm not loosing something on the line. 

Now imagine that I can successfully do that (which I doubt seriously) and I can decrypt a message in my browser, where can I save it to handle correctly the history ? The current Movim architecture is designed to save all the messages in the server SQL database ? So I'll create a second place to save only the "OMEMO/OTR" messages ? What if I also have unencrypted messages in between (that are coming from the SQL database)… If I send them back to the server SQL database, there's no point of doing this (because the Movim pod will have an unencrypted version), if I don't save them and if you reload your page, you will basically loose your history.

I know that OMEMO is solving most of the issues that OTR had (especially to handle E2E sessions using several XMPP clients at the same time), but it was always designed in a way that the encryption/decryption layer is done at the same place where the XMPP session is handled and where the messages are stored. Movim doesn't have this architecture (and it's not going to have it).

When I'm saying that it's not possible I don't want to tell you that I just don't want to do it because I'm lazy or because it's too difficult. I'd really like to have a nice and standard E2E encryption implementation in Movim like all of you. Just that I prefer to be clear about this point better than having an implementation that is not-secured because I had to do some comprises to do it.

Also, if you really want to do E2E encryption on your XMPP account, Movim doesn't forbid you to use another client to do it, that's the power of XMPP :p

P.S.: @inputmice correct me if I said something wrong or if I totally missed something about this topic

**Comment by louy2** [comment]
@edhelas 
I see Movim is not a conventional XMPP IM app, which is what excited me to check it out. But part of the reason I am excited about an XMPP SNS is because the E2E crypto XMPP provides may enable a better privacy structure. So as I totally understand the situation, I can't help but feel a bit sad.

That said Movim is still a great step towards that. Will definitely come for inspiration. Take care.

**Comment by rhz** [comment]
What about storing the messages encrypted in the server and using a second password as a master key to decode the encrypted messages in the client?

**Comment by edhelas** [comment]
I'm definitely closing this ticket. If you still want to talk about it, please join us on our chatroom : movim@conference.movim.eu.

**Comment by edhelas** [comment]
Reopening the ticket. I'll give more information about it soon.

**Comment by tcitworld** [comment]
So ?

**Comment by mrksr** [comment]
@edhelas said on the movim chatroom a few days ago, that he would post an update to this issue after finishing the next release, which according to him is almost done.

**Comment by Dragnucs** [comment]
@edhelas: i just want to point you to some direction even if i might be totally wrong. We all agree that decryption should be done client side using JavaScript or some browser plugin (JS is better) and not having to store decrypted messages server-side.

Let us assume that Movim is able deliver non-OMEMO message as-is to the browser which just displays them. Then come OMEMO message, encrypted and the server doesn't know what to do with them. Currently, Movim is able to get them to the right chat room and displays them as E2E messages and does not decrypt them. This behavior is key to my understanding, because this can be used to add an HTML class to those messages and then having a JS function/event that looks these particular classes and tries to decrypt them. This decryption would use a key stored in [LocalStorage](https://www.w3.org/TR/webstorage/#the-localstorage-attribute). Now each time Movim is unable to find this information, it needs to create a new one as a new client. In my personal opinion this is better security. Why? Because this means the user is connecting from a new location which Movim has no prior knowledge of, that  might be an unwanted connection, and thus, the user's contacts would have to allow a new key and be consious about the situation.

Using local storage would allow all tabs and windows to have the same key, while each tab would have to do the decryption on its own. This is not an issue because this would be on a computer (who opens many tabs of Movim on mobile?)

To summarize, each 'localStorage' is a 'client', and each tab looks for encrypted messages and decrypts them, in-place, regardless of what other tabs have done.

Users would see something like, i load a tab, see some non-OMEMO messages and a bunch of 'decryption in progress' stuff and after about 0.456 seconds, all the loading stuff would be transformed to real messages.

Once I was sitting with a friend using facebook with crappy connection, and basically, when you load your timeline, all you see is what is unchanged all over facebook, a textarea to type your status in and then, a bunch of place holders looking like user posts. After messages are loaded, placeholders are replaced with actual statuses.

**Comment by edhelas** [comment]
@Dragnucs, thanks you practically summed-up what I was going to say :) Yes the session, keys and messages will actually stay in localStorage and each Movim "browser" will be an OMEMO client by itself.

**Comment by ghost** [comment]
Would this still work for those who use privacy modes or otherwise empty out LocalStorage? How would everything know that the new login is a good login? 

`Because this means the user is connecting from a new location which Movim has no prior knowledge of, that might be an unwanted connection, and thus, the user's contacts would have to allow a new key and be consious about the situation.`

If I understand this right, does this mean that people who don't keep LocalStorage intact, would constantly have to have their contacts verify who they are? This seems like it'd be super cumbersome for the average user, and the repetition could easily bring in "thoughtless acceptance" -- much like how people mindlessly press "I Agree" on places with EULAs.

To me, this seems like there's very little safety in the event of an account breach.

**Comment by ghost** [comment]
It might be useful to look at how the people at Riot.im handle that: It's no problem for the electron app which remains always logged in. This should be the preferred way to run a privacy-sensitive web app anyway. People who insist on using their browser are warned before logging out and thus invalidating their LocalStorage and are given the chance to export their key material and to re-import it when they re-login, thus making re-verification unnecessary. Of course, if someone insists on removing all data that proves their cryptographic identity in some way, there is no way around revalidation.

**Comment by edhelas** [comment]
This feature will not be integrated in the upcoming release (0.12), also there is some serious discussions in the XSF regarding the standardization process of OMEMO, I'm waiting to see where it goes before taking a decision how it should be done as it could have quite a big impact on the result (and the current Conversations/Gajim implementation doesn't respect the current draft).

**Comment by edhelas** [comment]
I'm closing this feature request. I'm not planning to implement OMEMO or related E2E encryption in Movim, I already explained in this ticket details about why it was not a good idea. 

In short, Movim is a "light" client and relies heavily on the web server to handle, store, parse, analyze the messages. E2E encryption requires all those work to be done in the browser, this features goes against the architecture and Movim's model. Furthermore, even if that encryption was performed on the browser side, the code that handle that will be sent from the web server (which can be hacked…).

I really tried to find a way to handle that, but Movim is not made for OMEMO. Hopefully we still fully rely on XMPP, so you have the liberty to use another "local" client to send and receive encrypted messages next to Movim :)

**Comment by matlag** [comment]
I still think you should, and store the key right on the pod, with a big warning to the user that the key is stored on the pod and the admin can access it.
Use case:
-I have my own pod, that I fully trust, simply because that's my very own pod!
-I exchange with a user who's using a different client on a server he does not trust
We want to have a secure conversation.
I can still use Movim, with the benefit of not dealing with keys transfer from device to device.

**Comment by Dragnucs** [comment]
@edhelas What if we hire someone from Upwork to do this and give him the damn $100?

**Comment by alexandre1985** [comment]
Have someone considered writing a browser extension that would enable OTR and OMEMO on web clients? Can't this solve your problems?
Privacy is important for users.
And in fact if someone is swapping from facebook to movim this may exactly be the reason... Think about it

**Comment by Neustradamus** [comment]
important missing feature...

**Comment by Letterus** [comment]
You might have a look at https://crypto.cat - they are trying to get encryption over XMPP right - for a long time. It's a very hard job. You would at least need a bundled software for each platform, which integrity can be veryfied during update processes. Maintaining such a software bundle would be a project of its own, which requires high knowledge regarding cryptography…

**Comment by edhelas** [comment]
After having some though I'm reopening this ticket.
I'll try to add OMEMO in Movim, but it will be done "server side" in PHP (where the XMPP session is actually handled) because it's the only place where I can do it regarding the current architecture. Movim will clearly explain that particularity (that could give a false sense of security in some cases) through the UI.

This is not planned for the upcoming release 0.14 but the next one 0.15.

**Comment by edhelas** [comment]
For the moment I cannot start to work on this feature because of the PHP 7.3 related issue (https://github.com/movim/movim/issues/731) and the following missing PHP dependencies in the Debian repositories:
* ext-curve25519
* ext-protobuf
* ext-crypto

**Comment by ghost** [comment]
Any progress?

**Comment by Natureshadow** [comment]
> For the moment I cannot start to work on this feature because of the PHP 7.3 related issue (https://github.com/movim/movim/issues/731) and the following missing PHP dependencies in the repositories:
> * ext-curve25519
> * ext-protobuf
> * ext-crypto

Would it help if I built and uploaded them to Debian unstable?

**Comment by Natureshadow** [comment]
So, let's see:

- [x] protobuf: I just submitted a patch to build the PHP package in Debian, but it is also available from composer - PECL extension is not needed
- [ ] crypto: Do you mean bukka/php-crypto (an experimental OpenSSL binding library)?
- [ ] curve25519: This one? mgp25/curve25519-php

**Comment by edhelas** [comment]
Basically the idea would be to use https://github.com/mgp25/libsignal-php :)

**Comment by joe-average-user** [comment]
I would like to use movim as webclient for xmpp. Unfortunately most of the people I chat with use OMEMO. So at the current stage movim does not make a lot of sense for me. Nevertheless I read the issue and understand the authors' basic concerns of not really being end-to-end encryption based on the whole concept of movim.
Nevertheless I would like to speak up on the issue. I have years experience in ISP business and had to learn that most endusers have few - not to say no - control over their own PCs. Basically they click on every "confirm" button that pops up. Of course I don't say that it a good way to work on a home PC, but one has to understand that there is merely no knowledge about internet, ip, http, java, javascript and all the other things needed for using a PC today at all.
Which leads me to this statement:
dear author, please assume that the security on a webserver no matter who set it up, is for sure better than on any enduser PC with a local end-to-end encrypting app. And don't forget that the ones not being satisfied by a webclient with its drawbacks can still ommit it and use their favourite client software. That's the good thing about an open standard like XMPP. 
I can assure you there is currently no web client for XMPP with useable OMEMO encryption available, and it would be great if you can movim the first one.
All infos on omemo.top about web clients supporting it are plain wrong. As an example lets name converse.js which does omemo on text messages but not on any binaries sent. And believe me, it won't in the future. That's not a useable implementation for standard users. Most of them have seen the pretty astonishing Conversations app (android) and are not content with any less.
Hope you can do it!
Thanks for your work

**Comment by Natureshadow** [comment]
I recently was reminded that my backlog of work is blocking this issue due to lack of some PHP modules in Debian.

I took up work on it again now.

**Comment by joe-average-user** [comment]
How are things going in this issue?

**Comment by deuill** [comment]
Is there a public branch we can follow? I can help push this over the edge, if there's a clear plan for the implementation.

**Comment by edhelas** [comment]
I'm still waiting for the dependencies to be there to start to work on it :)

**Comment by joe-average-user** [comment]
Can you name what is missing? I try to look after that then ...

**Comment by Natureshadow** [comment]
> Can you name what is missing? I try to look after that then ...

Please learn to read. It is all explained here.

Also, I am working on it. Do not duplicate work.

**Comment by herbsmn** [comment]
@Natureshadow you are being rude and unhelpful.

**Comment by edhelas** [comment]
@Natureshadow it seems that there's another way to add OMEMO within Movim.

Since PHP 7.4 (that was released a few days ago) you can map C libraries to PHP (https://www.php.net/manual/en/book.ffi.php). It would then be simple for me to map the C libsignal library (that is already in the repos) to PHP and simply use it in Movim.

It's still some big guesses that I'm making, but I'll have a look as soon as PHP 7.4 lands in the testing repos ;)

Cheers

**Comment by Natureshadow** [comment]
Yeah, that's basically what was one of the open questions, how much needs to be ported nowadays.

**Comment by apimon** [comment]
It is totally absurd talking about end-to-end encryption and doing it "on the server". 
Dangerous nonsense.

**Comment by edhelas** [comment]
Wow damn, guys, lets STOP EVERYTHING. Looks like we overlooked that E2EE on Movim will have to be done "on the server".
![](http://www.commitstrip.com/wp-content/uploads/2016/06/Strip-Commentaires-sur-Mars-650-finalenglish.jpg)

I'm glad that we noticed that after 5 years of discussion in this open ticket.

**Comment by joe-average-user** [comment]
It's really hard to not call him an idiot. Nice cartoon nevertheless :-)

**Comment by edhelas** [comment]
@joe-average-user I'd prefer to not give bad names to eachothers :p 
Whatever, let's go back to this OMEMO topic, shall we :D ?

**Comment by Natureshadow** [comment]
OMEMO or Oh-Meme-Oh :D?

**Comment by joe-average-user** [comment]
@edhelas well, I'd love to, but after 4 completed years of no real action and after more or less being told to shut up by @Natureshadow because "he is working on it" I have very low expectations. Maybe we have quite a different method in software development (which I do for about 35 years now). In my company no code after 3 months means a project is plain dead and is dropped.

**Comment by Neustradamus** [comment]
It is like Duke Nukem Forever ah ah

Solution: There are other XMPP web intefaces with OMEMO :)

**Comment by Natureshadow** [comment]
> In my company no code after 3 months means a
>project is plain dead and is dropped.

Can you do me a favour and ask one of your company representatives what the average salary of their employees is? If they agree that $100 matches approx. 6 full-time days of developer work, I would very much like to see both your developers and your company representative for a group photo. If they don't, please either stop such comparisons or get @edhelas and me those 6 full-time days' salary and we will most certainly get OMEMO ready before end of the year.

In any case, I did not tell you to shut up. I asked you to not interfere *with the Debian packaging of PHP dependencies*. This is somewhere near 10% of the efforts of this project, and if you have the resources to do the remaining 90%, noone stops you from anything and I am looking very much forward to your timely implementation.

**Comment by joe-average-user** [comment]
In times of EU GDPR asking publicly for a photo is somewhat odd :-)
Nevertheless, debian is not the only distribution available. If you want to use the latest versions of something maybe you are better off with a rolling release distro like arch linux anyway?

@Neustradamus: how about posting a hint to these/this other web interface(s)?

**Comment by Natureshadow** [comment]
>In times of EU GDPR asking publicly for a photo is somewhat odd :-)

How does GDPR have any opinion on someone asking you for a meeting to take a photo, and how does it have an opinion on photos whatsoever if taken for private purposes :-)? (I know you are joking, but the inflatious abuse of GDPR as a buzzword somewhat spoils such jokes)

>Nevertheless, debian is not the only distribution available. If you
>want to use the latest versions of something maybe you are better off
>with a rolling release distro like arch linux anyway?

Only that we are talking about Debian sid, which is more or less a rolling release distro, and neither Arch nor Gentoo having the required packages :D.

**Comment by joe-average-user** [comment]
Iff the stuff you want to use already exists, why don't you compile it yourself (which is more or less anyway what e.g. gentoo does) and implement the movim features, so that when someday your packages are available for the distros movim can already use it? This way there is no time lost waiting for Godot ...

**Comment by Dragnucs** [comment]
There are also PECL packages, Docker images, snaps, flatpaks, PPAs, COPR, AUR, other distributions. Why should we be limited only yo what Debian provides?

**Comment by Natureshadow** [comment]
I am not convinced the discussion about whys will lead anywhere. I can assure you that Movim lacking OMEMO support is a huge issue for me as well, and both @edhelas and I are working on it with the best efforts we can (while also waiting for your PR, of course - again, if you think you can do it now, please do it.)

**Comment by Dragnucs** [comment]
@Natureshadow I cannot do it and donnot have enough time to learn how to do it. XMPP has its challenges.

I, like many others, switched to other thechnologies supporting end to end encryption, matrix for instance.

For now, I am just trying to discover solutions.

**Comment by maranda** [comment]
> In my company no code after 3 months means a project is plain dead and is dropped.
> Can you do me a favour and ask one of your company representatives what the average salary of their employees is? If they agree that $100 matches approx. 6 full-time days of developer work, I would very much like to see both your developers and your company representative for a group photo. If they don't, please either stop such comparisons or get @edhelas and me those 6 full-time days' salary and we will most certainly get OMEMO ready before end of the year. In any case, I did not tell you to shut up. I asked you to not interfere *with the Debian packaging of PHP dependencies*. This is somewhere near 10% of the efforts of this project, and if you have the resources to do the remaining 90%, noone stops you from anything and I am looking very much forward to your timely implementation.

Just read the issue, and even if I'm not sure if that was irony or not but 100$ is, maybe, 1 billed hour of freelance full-dev time not nearly one week..

To arrive straight to the point instead of complaining the people interested should:
1) Invest money and sponsor the work with the appropriate wage (which for example for a week at 4h/d-24h/w work would be 2400$ and not 100..)
2) Contribute yourself
3) Wait patiently

Because what is plainly misunderstood with FOSS, is that while the software might be free, the time of people behind that is not and they basicly donating that time to you.
So you either pick what they give you and be calm and quiet about it, or... That time or the eventual surplus of time needs to be compensated adeguately.

Just a thought.

**Comment by edhelas** [comment]
I'm pushing an early OMEMO implementation branch. With the financial help of [NLNet](https://nlnet.nl/)  and a specific architecture it seems that OMEMO is finally coming to Movim ;) I was able to send and receive encrypted message with other XMPP clients without doing any encryption server side.

More information soon!

**Comment by Neustradamus** [comment]
Good news, thanks NLnet!
- https://nlnet.nl/project/Movim-OMEMO

**Comment by edhelas** [comment]
I merged the https://github.com/movim/movim/pull/983 branch. It's an early working version, there is still bugs and weird behaviors but the one-to-one OMEMO chat works :)
I'll keep this ticket open until I have something stable and complete enough to close it.

So now, let's
![](https://www.marcus-povey.co.uk/wp-content/6a0120a85dcdae970b016301e98de2970d-800wi.png)

**Comment by edhelas** [comment]
As OMEMO is now part of Movim master and that the main working features are working I'm considering this ticket as solved.
There is still a lot of small things to fix but the main is done.

It took 6 years but Movim is finally having end to end encryption.

![](https://i.kym-cdn.com/entries/icons/original/000/029/768/Screen_Shot_2019-05-20_at_4.18.22_PM.png)

**Comment by Neustradamus** [comment]
OMEMO has been added in Movim 0.20!

Announcements:
- Movim 0.20 announcement -> https://mov.im/?node/pubsub.movim.eu/Movim/dd90cd1c-4bcd-45a5-88a3-220fd4280b21
- Movim 0.20 announcement -> HN: https://news.ycombinator.com/item?id=30475406
- Movim 0.20 announcement -> Movim subreddit: https://www.reddit.com/r/movim/comments/t1p7nz/movim_020_has_been_released_with_omemo_endtoend/
- Movim 0.20 announcement -> XMPP subreddit: https://www.reddit.com/r/xmpp/comments/t1p7tc/movim_020_has_been_released_with_omemo_endtoend/
